src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -1,21 +1,464 @@
+  only storing min-max information for key /prefixes/.
+
+  However, a /clash/ might still occur, albeit with a low probability. In a
+  clash, the boundary between pages is unclear because the prefix of the
+  maximum key for a page @i@ matches the prefix of the minimum key for a
+  successor page @i+1@. In this case, we will have to look at larger
+  bit-prefixes (or possibly even full keys) to determine the boundary between
+  pages. As such, for each clash we store more bits of data, but again this
+  only occurs with a low probability.
+
+  === Internals
+
+  TODO: explain range-finder bits, primary bits, secondary bits
+
+  @
+      Range-finder bits              Primary bits                  Secondary bits
+    |-------------------|---------------------------------------|------------------- .... {- elided -}
+     0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 .... {- elided -}
+  @
+-}
+
+-- | A compact fence-pointer index for uniformly distributed keys.
+--
+-- Compact indexes save space by storing bit-prefixes of keys.
+--
+data CompactIndex = CompactIndex {
+    -- | A vector for finding the range of 'ciPrimary' to search based on the
+    -- range-finder bits.
+    --
+    -- TODO: elaborate.
+    ciRangeFinder          :: !(VU.Vector Word32)
+    -- | Determines the size of 'ciRangeFinder' as @2 ^ 'ciRangeFinderPrecision'
+    -- + 1@.
+  , ciRangeFinderPrecision :: !Int
+
+    -- | Maps a page @i@ to the 32-bit slice of primary bits of its minimum key.
+  , ciPrimary              :: !(VU.Vector Word32)
+
+    -- | A clash on page @i@ means that the minimum key of page @i@ and the
+    -- maximum key of the predecessor page @i-1@ have the exact same @32@
+    -- primary bits. Use the tie-breaker index to break ties.
+  , ciClashes              :: !(VU.Vector Bit)
+    -- | Maps a page @i@ to its full minimum key, but only if there is a clash
+    -- on page @i@.
+    --
+    -- INVARIANT: a key is included in this 'IntMap' if and only if 'ciClashes'
+    -- recorded that there is a class. In code: @isJust (IntMap.lookup ix
+    -- ciTiebreaker) == ciClashes VU.! ix@.
+  , ciTieBreaker           :: !(IntMap SerialisedKey)
+  }
+
+{-------------------------------------------------------------------------------
+  Invariants
+-------------------------------------------------------------------------------}
+
+-- | Inclusive bounds for range-finder bit-precision.
+rangeFinderPrecisionBounds :: (Int, Int)
+rangeFinderPrecisionBounds = (0, 16)
+
+-- | https://en.wikipedia.org/wiki/Birthday_problem#Probability_of_a_shared_birthday_(collision)
+suggestRangeFinderPrecision :: Int -> Int
+suggestRangeFinderPrecision maxPages =
+    clamp $ ceiling @Double $ 2 * logBase 2 (fromIntegral maxPages)
+  where
+    (lb, ub)            = rangeFinderPrecisionBounds
+    clamp x | x < lb    = lb
+            | x > ub    = ub
+            | otherwise = x
```

Comment:
The clamping is wrong. The 2 * log_2 bit is right, but that gives the _total_ number of bits, e.g. for our example of 100e6/40 it'll give us 43, but then we need to clamp it to the range of 32..48 and then subtract 32 to give the bits above 32 that we need in the range finder bits.
```suggestion
    (lb, ub)            = rangeFinderPrecisionBounds
    clamp x | x < lb+32 = lb
            | x > ub+32 = ub
            | otherwise = x - 32
```
so then for example we get
```
suggestRangeFinderPrecision (100 * 10^6 `div` 40 :: Int)
11
```
whereas previously it'd give us silly results like
```
suggestRangeFinderPrecision (16 :: Int)
8
```
we don't want 8 bits of precision for 16 entries!